panda_deburring/ocp_croco_force_feedback.py

```diff
@@ -386,199 +406,115 @@ def create_terminal_model(self) -> crocoddyl.ActionModelAbstract:
+
+            # By using `__dict__["element"]` obtain references to functions in order to cache them
+            state_reg = running_model_diff.costs.costs["stateReg"]
+            ocp_ref_cache.state_reg = OCPReferences(
+                reference=state_reg.cost.residual.__dict__["reference"],
+                weights=state_reg.cost.activation.__dict__["weights"],
+            )
+
+            frame_translation = running_model_diff.costs.costs["frameTranslationCost"]
+            ocp_ref_cache.frame_translation = OCPReferences(
+                reference=frame_translation.cost.residual.__dict__["reference"],
+                weights=frame_translation.cost.activation.__dict__["weights"],
+                frame_id=frame_translation.cost.residual.__dict__["id"],
+            )
+
+            frame_rotation = running_model_diff.costs.costs["frameRotationCost"]
+            ocp_ref_cache.frame_rotation = OCPReferences(
+                reference=frame_rotation.cost.residual.__dict__["reference"],
+                weights=frame_rotation.cost.activation.__dict__["weights"],
+                frame_id=frame_rotation.cost.residual.__dict__["id"],
+            )
+
+            frame_velocity = running_model_diff.costs.costs["frameVelocityCost"]
+            ocp_ref_cache.frame_velocity = OCPReferences(
+                reference=frame_velocity.cost.residual.__dict__["reference"],
+                weights=frame_velocity.cost.activation.__dict__["weights"],
+                frame_id=frame_velocity.cost.residual.__dict__["id"],
+            )
+
+            ocp_ref_cache.control_reg = OCPReferences(
+                reference=None,
+                weights=running_model_diff.__dict__["tau_grav_weight"],
+            )
+
+            ocp_ref_cache.force_tracking = OCPReferences(
+                use_residual=running_model_diff.__dict__["active_contact"],
+                reference=running_model_diff.__dict__["f_des"],
+                weights=running_model_diff.__dict__["f_weight"],
+            )
+
+            self._references_cache.append(ocp_ref_cache)
+
     def set_reference_weighted_trajectory(
         self, reference_weighted_trajectory: list[WeightedTrajectoryPoint]
     ):
         """Set the reference trajectory for the OCP."""
         assert len(reference_weighted_trajectory) == self.n_controls + 1
 
         # Modify running costs reference and weights
-        for i, ref_weighted_pt in enumerate(reference_weighted_trajectory[:-1]):
-            # Modifying the state regularization cost
-            state_reg = self._solver.problem.runningModels[i].differential.costs.costs[
-                "stateReg"
-            ]
-            state_reg.cost.residual.reference = np.concatenate(
+        for i in range(len(reference_weighted_trajectory)):
+            ref_weighted_pt = reference_weighted_trajectory[i]
+            ocp_ref = self._references_cache[i]
+
+            ocp_ref.state_reg.reference = np.concatenate(
```

Comment:
You can replace this `concatenate` by smt like
```python
ocp_ref.state_reg.reference[:n] = ref_weighted_pt.point.robot_configuration
ocp_ref.state_reg.reference[n:] = ref_weighted_pt.point.robot_velocity
```
Note that I'm not entirely sure that this works as it does not go through the setter of attribute `reference`. I had the issue with crocoddyl in the past but I don't remember for which cost.

Also, I think the attribute `robot_state` exists.